### PR TITLE
KAFKA-17270: Let test code use Exit.java rather than Exit.scala

### PR DESCRIPTION
**Pull request by frankvicky**
JIRA: [KAFKA-17270](https://issues.apache.org/jira/browse/KAFKA-17270)

`Exit.scala` is a wrapper around `Exit.java`, and its main benefit is that it allows methods to throw exceptions by returning `Nothing`. This benefit is minimal, and since we are planning to phase out Scala code, it would be better to use `Exit.java` directly and remove `Exit.scala`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
It seems `Exit.scala` is not removed by this PR?

**Comment by frankvicky** [comment]
It seems that I have a misunderstanding.
I will remove Exit.scala entirely in next commit.😃

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 1]
```diff
@@ -6701,7 +6702,7 @@ class ReplicaManagerTest {
       def haltProcedure(exitStatus: Int, message: Option[String]): Nothing = {
         fail("Test failure, broker should not have halted")
       }
-      Exit.setHaltProcedure(haltProcedure)
+      Exit.setHaltProcedure((status, message) => haltProcedure(status, Option(message)))
```
Comment (gharris1727) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: this can be inlined, now that it requires a lambda wrapper. The arguments can be `(_, _)` because the body doesn't depend on their values.

File: core/src/test/scala/kafka/utils/ExitTest.scala [hunk 2]
```diff
@@ -18,6 +18,7 @@
 package kafka.utils
 
 import java.io.IOException
+import org.apache.kafka.common.utils.Exit
```
Comment (gharris1727) [inline core/src/test/scala/kafka/utils/ExitTest.scala]: There's already a Java ExitTest which is identical, so this can be removed rather than migrated.

**Review by gharris1727 — COMMENTED** [review]
Thanks @frankvicky for the PR!

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 1 above)
Comment (frankvicky) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Cool, it looks more concise, learn a trick 😺 
Thank you

**Comment by frankvicky** [comment]
Hi @gharris1727 , @chia7712 
I have entirely removed `Exit.scala` and replaced it with `Exit.java`, PTAL
Many Thanks 🐧

File: core/src/main/scala/kafka/Kafka.scala [hunk 3]
```diff
@@ -21,7 +21,8 @@ import java.util.Properties
 import joptsimple.OptionParser
 import kafka.server.{KafkaConfig, KafkaRaftServer, KafkaServer, Server}
 import kafka.utils.Implicits._
-import kafka.utils.{Exit, Logging}
+import kafka.utils.Logging
+import org.apache.kafka.common.utils.Exit
```
Comment (mimaison) [inline core/src/main/scala/kafka/Kafka.scala]: You can merge this with the other `org.apache.kafka.common.utils` imports below.

Same in other files.

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 4]
```diff
@@ -22,6 +22,7 @@ import joptsimple._
 import joptsimple.util.EnumConverter
 import kafka.security.authorizer.AclAuthorizer
 import kafka.utils._
+import org.apache.kafka.common.utils.Exit
```
Comment (mimaison) [inline core/src/main/scala/kafka/admin/AclCommand.scala]: Can we keep the imports ordered?
Same in other files below

File: core/src/main/scala/kafka/docker/KafkaDockerWrapper.scala [hunk 5]
```diff
@@ -18,13 +18,14 @@ package kafka.docker
...
-import kafka.utils.{Exit, Logging}
+import kafka.utils.Logging
 import net.sourceforge.argparse4j.ArgumentParsers
 import net.sourceforge.argparse4j.impl.Arguments.store
 import net.sourceforge.argparse4j.inf.Namespace
+import org.apache.kafka.common.utils.Exit
 
 import java.nio.charset.StandardCharsets
-import java.nio.file.{Files, Path, Paths, StandardCopyOption, StandardOpenOption}
+import java.nio.file._
```
Comment (mimaison) [inline core/src/main/scala/kafka/docker/KafkaDockerWrapper.scala]: Can we undo this unrelated change?

File: core/src/main/scala/kafka/tools/TestRaftServer.scala [hunk 6]
```diff
@@ -33,18 +30,21 @@ import org.apache.kafka.common.metrics.stats.{Meter, Percentile, Percentiles}
 import org.apache.kafka.common.protocol.{ObjectSerializationCache, Writable}
 import org.apache.kafka.common.security.scram.internals.ScramMechanism
 import org.apache.kafka.common.security.token.delegation.internals.DelegationTokenCache
-import org.apache.kafka.common.utils.{Time, Utils}
+import org.apache.kafka.common.utils.{Exit, Time, Utils}
 import org.apache.kafka.common.{TopicPartition, Uuid, protocol}
 import org.apache.kafka.raft.errors.NotLeaderException
-import org.apache.kafka.raft.{Batch, BatchReader, Endpoints, LeaderAndEpoch, RaftClient, QuorumConfig}
+import org.apache.kafka.raft._
```
Comment (mimaison) [inline core/src/main/scala/kafka/tools/TestRaftServer.scala]: Let's undo this unrelated change

**Comment by frankvicky** [comment]
Hi @mimaison 
I have reordered the imports, PTAL
Many Thanks 😺

File: core/src/main/scala/kafka/Kafka.scala (hunk 3 above)
Comment (gharris1727) [reply core/src/main/scala/kafka/Kafka.scala]: I think this one is still unresolved.

Comment (frankvicky) [reply core/src/main/scala/kafka/Kafka.scala]: I will take a look

File: core/src/main/scala/kafka/admin/AclCommand.scala (hunk 4 above)
Comment (gharris1727) [reply core/src/main/scala/kafka/admin/AclCommand.scala]: @frankvicky How did you reorder the imports? The diff looks a lot larger now.

IMHO we can just insert the changed imports "close enough" to the correct order, and leave most of this as-is. Separately we could try and apply spotless in scala, (or leave it as-is until the scala gets removed)

@mimaison what did you have in mind here?

Comment (frankvicky) [reply core/src/main/scala/kafka/admin/AclCommand.scala]: Hi @gharris1727 
I have leveraged the function from IDEA while preserving the original imports if they were using specific import. If it was a wildcard import, I left it as it was. 👀

File: core/src/main/scala/kafka/Kafka.scala (hunk 3 above)
Comment (frankvicky) [reply core/src/main/scala/kafka/Kafka.scala]: It should be fixed now. Sorry for missing the file. 😥

File: core/src/main/scala/kafka/admin/AclCommand.scala (hunk 4 above)
Comment (mimaison) [reply core/src/main/scala/kafka/admin/AclCommand.scala]: I don't think we should reorder all imports, that just creates noise.
But here let's add the import with the other `import org.apache.kafka.common.utils` imports we have below instead of inserting at a pretty random place.

Comment (frankvicky) [reply core/src/main/scala/kafka/admin/AclCommand.scala]: I see.
I will do it in next commit 😃

File: core/src/test/scala/kafka/utils/ExitTest.scala (hunk 2 above)
Comment (chia7712) [reply core/src/test/scala/kafka/utils/ExitTest.scala]: agree to remove `ExitTest.scala`. 

unresolve this conversation as this comment is not addressed. @frankvicky Please update it in next comment, thanks!

File: core/src/main/scala/kafka/admin/AclCommand.scala (hunk 4 above)
Comment (chia7712) [reply core/src/main/scala/kafka/admin/AclCommand.scala]: There are many import chaos in scala code, and that is why I do thank @gongxuanzhang to complete the old but important [KAFKA-10787](https://issues.apache.org/jira/browse/KAFKA-10787) which reducing the import conflicts of java code in the future.

File: core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala [hunk 7]
```diff
@@ -34,9 +34,9 @@ import org.apache.zookeeper.data.Stat
 
 import scala.annotation.tailrec
 import scala.collection.mutable
-import scala.jdk.CollectionConverters._
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala]: please revert this change

**Review by chia7712 — COMMENTED** [review]
@frankvicky Thanks for this patch, but could you please add the import of `Exit.java` with less import changes? Those scala code has no import formatter, and our IDE could have different import order.

**Comment by frankvicky** [comment]
Hi everyone,
I have addressed all comments. PTAL
Many thanks for your patience.

**Review by chia7712 — APPROVED** [review]
LGTM. will merge it if other reviews have no objection.

**Comment by gharris1727** [comment]
Thanks @frankvicky @chia7712 @mimaison ! Good to see this simplification.
